File: AEDS/AEDS2/src/beecrowds/camp/c/camp.c
#include <stdio.h>
#include <stdlib.h>

typedef struct Crianca
{
  char *nome;
  int e;
} Crianca;

typedef struct Celula
{
  Crianca c;
  struct Celula *prox, *ant;
} Celula;

Celula *novaCelula(int x)
{
  Celula *nova = malloc(sizeof(Celula));
  nova->prox = NULL;
  nova->ant = NULL;
  nova->e = x;
  return nova;
};

int tam(Celula *cabeca)
{
  int tam = 1;
  Celula *atual = cabeca->prox;
  for (; atual != NULL; atual = atual->prox, tam++)
    ;
  return tam;
}

void inserir(Celula *cabeca, int x)
{
  Celula *nova = novaCelula(x);
  Celula *atual = cabeca;
  while (atual->prox != cabeca)
  {
    atual = atual->prox;
  }
  atual->prox = nova;
  nova->ant = atual;
  nova->prox = cabeca;
};

int remover(Celula **cabeca, int passos)
{
  if (*cabeca == NULL)
    return -1;

  Celula *atual = (passos % 2 == 0) ? caminhaDir(*cabeca, passos)
                                    : caminhaEsq(*cabeca, passos);

  if (atual == *cabeca)
  {
    if (atual->prox == atual)
    {
      int e = atual->e;
      free(atual);
      *cabeca = NULL;
      return e;
    }
    atual->ant->prox = atual->prox;
    atual->prox->ant = atual->ant;
    *cabeca = atual->prox;
    int e = atual->e;
    free(atual);
    return e;
  }

  atual->ant->prox = atual->prox;
  atual->prox->ant = atual->ant;
  int e = atual->e;
  free(atual);
  return e;
}

Celula *caminhaDir(Celula *cabeca, int passos)
{
  Celula *tmp = cabeca;
  for (int i = 0; i < passos; i++)
    tmp = tmp->prox;
  return tmp;
}

Celula *caminhaEsq(Celula *cabeca, int passos)
{
  Celula *tmp = cabeca;
  for (int i = 0; i < passos; i++)
    tmp = tmp->ant;
  return tmp;
}

void limpar(Celula *cabeca)
{
  Celula *atual = cabeca->prox;
  Celula *libera;
  while (atual != NULL)
  {
    libera = atual;
    atual = atual->prox;
    free(libera);
  }
};

int main()
{
  Celula *cabeca = novaCelula(0);

  inserir(cabeca, 1);
  inserir(cabeca, 2);

  imprime(cabeca);

  limpar(cabeca);
  return 0;
}

File: AEDS/AEDS2/src/beecrowds/camp/java/Camp.java
package beecrowds.camp.java;

import java.util.Scanner;

public class Camp {
	public static class Crianca {
		String nome;
		int ficha;
		Crianca ant, prox;

		public Crianca(String nome, int ficha) {
			this.nome = nome;
			this.ficha = ficha;
			this.ant = this.prox = null;
		}

	}

	public static class Lista {
		Crianca inicio, fim;

		public Lista() {
			this.fim = null;
			this.inicio = fim;
		}

		public void inserir(Crianca c) {
			if (inicio == null) {
				inicio = fim = c;
				c.prox = c;
				c.ant = c;
				return;
			}
			fim.prox = c;
			c.ant = fim;
			fim = c;
			c.prox = inicio;
			inicio.ant = c;
		}

		public Crianca remover(Crianca comeco, int passos) {
			Crianca tmp = caminha(comeco, passos);
			if (inicio == tmp.prox) {
				int e = inicio.ficha;
				inicio = inicio.prox;
				fim.prox = inicio;
				inicio.ant = fim;
				return e;
			} else if (fim == tmp.prox) {
				Crianca e = fim.ficha;
				fim = fim.ant;
				inicio.ant = fim;
				fim.prox = inicio;
				return e;
			}
			Crianca e = tmp.prox.ficha;
			tmp.prox = tmp.prox.prox;
			tmp.prox.ant = tmp;
			return e;
		}

		public Crianca caminha(Crianca comeco, int passos) {
			Crianca tmp = comeco;
			if (passos % 2 == 0) {
				for (int i = 0; i < passos; i++, tmp = tmp.prox)
					;
			} else {
				for (int i = 0; i < passos; i++, tmp = tmp.ant)
					;
			}
			return tmp;
		}

		public String vencedor() {
			if (inicio == fim && inicio != null) {
				return inicio.nome;
			}
			return "deu merda";
		}

		public void imprime() {
			Crianca tmp = inicio;
			System.out.printf("[%s ", tmp.nome);
			tmp = tmp.prox;
			for (; tmp != inicio; tmp = tmp.prox)
				System.out.printf("%s ", tmp.nome);

			System.out.printf("]\n");
		}

	}

	public static void jogo(Lista lista, int n) {
		int passos = lista.inicio.ficha;
		while (n > 1) {
			passos = lista.remover(passos);
			n--;
			lista.imprime();
		}
		System.out.println("Vencedor(a): " + lista.vencedor());
	}

	public static void main(String[] args) {
		Scanner scanf = new Scanner(System.in);
		String nCriancas = scanf.nextLine();
		int n = Integer.parseInt(nCriancas);
		while (n > 0) {
			Lista lista = new Lista();
			for (int i = 0; i < n; i++) {
				String entrada = scanf.nextLine();

				String[] splitted = entrada.split(" ");
				lista.inserir(new Crianca(splitted[0], Integer.parseInt(splitted[1])));
			}

			jogo(lista, n);

			n = Integer.parseInt(scanf.nextLine());
		}
	}
}
